File: comscore/android/src/main/java/com/theoplayercomscore/integration/ComscoreMetaData.java
package com.theoplayercomscore.integration;

import androidx.annotation.NonNull;

import com.comscore.streaming.ContentMetadata;

import java.util.Map;

public class ComscoreMetaData {

  private ComscoreMediaType mediaType;
  private String uniqueId;
  private long length;
  private String c3 = null;
  private String c4 = null;
  private String c6 = null;
  private String stationTitle;
  private String stationCode = null;
  private String networkAffiliate = null;
  private String publisherName = null;
  private String programTitle;
  private String programId = null;
  private String episodeTitle;
  private String episodeId = null;
  private String episodeSeasonNumber = null;
  private String episodeNumber = null;
  private String genreName;
  private String genreId = null;
  private Boolean carryTvAdvertisementLoad = null;
  private Boolean classifyAsCompleteEpisode = null;
  private ComscoreDate dateOfProduction = null;
  private ComscoreTime timeOfProduction = null;
  private ComscoreDate dateOfTvAiring = null;
  private ComscoreTime timeOfTvAiring = null;
  private ComscoreDate dateOfDigitalAiring = null;
  private ComscoreTime timeOfDigitalAiring = null;
  private ComscoreFeedType feedType = null;
  private Boolean classifyAsAudioStream;
  private ComscoreDeliveryMode deliveryMode = null;
  private ComscoreDeliverySubscriptionType deliverySubscriptionType = null;
  private ComscoreDeliveryComposition deliveryComposition = null;
  private ComscoreDeliveryAdvertisementCapability deliveryAdvertisementCapability = null;
  private ComscoreMediaFormat mediaFormat = null;
  private ComscoreDistributionModel distributionModel = null;
  private String playlistTitle = null;
  private Integer totalSegments = null;
  private String clipUrl = null;
  private ComscoreDimension videoDimension = null;
  private Map<String,String> customLabels = null;

    public ComscoreMetaData(
      ComscoreMediaType mediaType,
      String uniqueId,
      long length,
      String c3,
      String c4,
      String c6,
      String stationTitle,
      String stationCode,
      String networkAffiliate,
      String publisherName,
      String programTitle,
      String programId,
      String episodeTitle,
      String episodeId,
      String episodeSeasonNumber,
      String episodeNumber,
      String genreName,
      String genreId,
      Boolean carryTvAdvertisementLoad,
      Boolean classifyAsCompleteEpisode,
      ComscoreDate dateOfProduction,
      ComscoreTime timeOfProduction,
      ComscoreDate dateOfTvAiring,
      ComscoreTime timeOfTvAiring,
      ComscoreDate dateOfDigitalAiring,
      ComscoreTime timeOfDigitalAiring,
      ComscoreFeedType feedType,
      Boolean classifyAsAudioStream,
      ComscoreDeliveryMode deliveryMode,
      ComscoreDeliverySubscriptionType deliverySubscriptionType,
      ComscoreDeliveryComposition deliveryComposition,
      ComscoreDeliveryAdvertisementCapability deliveryAdvertisementCapability,
      ComscoreMediaFormat mediaFormat,
      ComscoreDistributionModel distributionModel,
      String playlistTitle,
      Integer totalSegments,
      String clipUrl,
      ComscoreDimension videoDimension,
      Map<String,String> customLabels
    ) {
      this.mediaType = mediaType;
      this.uniqueId = uniqueId;
      this.length = length;
      this.c3 = c3;
      this.c4 = c4;
      this.c6 = c6;
      this.stationTitle = stationTitle;
      this.stationCode = stationCode;
      this.networkAffiliate = networkAffiliate;
      this.publisherName = publisherName;
      this.programTitle = programTitle;
      this.programId = programId;
      this.episodeTitle = episodeTitle;
      this.episodeId = episodeId;
      this.episodeSeasonNumber = episodeSeasonNumber;
      this.episodeNumber = episodeNumber;
      this.genreName = genreName;
      this.genreId = genreId;
      this.carryTvAdvertisementLoad = carryTvAdvertisementLoad;
      this.classifyAsCompleteEpisode = classifyAsCompleteEpisode;
      this.dateOfProduction = dateOfProduction;
      this.timeOfProduction = timeOfProduction;
      this.dateOfTvAiring = dateOfTvAiring;
      this.timeOfTvAiring = timeOfTvAiring;
      this.dateOfDigitalAiring = dateOfDigitalAiring;
      this.timeOfDigitalAiring = timeOfDigitalAiring;
      this.feedType = feedType;
      this.classifyAsAudioStream = classifyAsAudioStream;
      this.deliveryMode = deliveryMode;
      this.deliverySubscriptionType = deliverySubscriptionType;
      this.deliveryComposition = deliveryComposition;
      this.deliveryAdvertisementCapability = deliveryAdvertisementCapability;
      this.mediaFormat = mediaFormat;
      this.distributionModel = distributionModel;
      this.playlistTitle = playlistTitle;
      this.totalSegments = totalSegments;
      this.clipUrl = clipUrl;
      this.videoDimension = videoDimension;
      this.customLabels = customLabels;
    }

    public ContentMetadata toComscoreContentMetadata() {
      ContentMetadata.Builder cm = new ContentMetadata.Builder();
      cm.mediaType(mediaType.toComscore());
      cm.uniqueId(uniqueId);
      cm.length(length);
      if(c3 != null) cm.dictionaryClassificationC3(c3);
      if(c4 != null) cm.dictionaryClassificationC4(c4);
      if(c6 != null) cm.dictionaryClassificationC6(c6);
      cm.stationTitle(stationTitle);
      if(stationCode != null) cm.stationCode(stationCode);
      if(networkAffiliate != null) cm.networkAffiliate(networkAffiliate);
      if(publisherName != null) cm.publisherName(publisherName);
      cm.programTitle(programTitle);
      if(programId != null) cm.programId(programId);
      cm.episodeTitle(episodeTitle);
      if(episodeId != null) cm.episodeId(episodeId);
      if(episodeSeasonNumber != null) cm.episodeSeasonNumber(episodeSeasonNumber);
      if(episodeNumber != null) cm.episodeNumber(episodeNumber);
      cm.genreName(genreName);
      if(genreId != null) cm.genreId(genreId);
      if(carryTvAdvertisementLoad != null) cm.carryTvAdvertisementLoad(carryTvAdvertisementLoad);
      if(classifyAsCompleteEpisode != null) cm.classifyAsCompleteEpisode(classifyAsCompleteEpisode);
      if(dateOfProduction != null) cm.dateOfProduction(dateOfProduction.getYear(),dateOfProduction.getMonth(),dateOfProduction.getDay());
      if(timeOfProduction != null) cm.timeOfProduction(timeOfProduction.getHours(),timeOfProduction.getMinutes());
      if(dateOfTvAiring != null) cm.dateOfTvAiring(dateOfTvAiring.getYear(),dateOfTvAiring.getMonth(),dateOfTvAiring.getDay());
      if(timeOfTvAiring != null) cm.timeOfTvAiring(timeOfTvAiring.getHours(),timeOfTvAiring.getMinutes());
      if(dateOfDigitalAiring != null) cm.dateOfDigitalAiring(dateOfDigitalAiring.getYear(),dateOfDigitalAiring.getMonth(),dateOfDigitalAiring.getDay());
      if(timeOfDigitalAiring != null) cm.timeOfDigitalAiring(timeOfDigitalAiring.getHours(),timeOfDigitalAiring.getMinutes());
      if(feedType != null) cm.feedType(feedType.toComscore());
      cm.classifyAsAudioStream(classifyAsAudioStream);
      if(deliveryMode != null) cm.deliveryMode(deliveryMode.toComscore());
      if(deliverySubscriptionType != null) cm.deliverySubscriptionType(deliverySubscriptionType.toComscore());
      if(deliveryComposition != null) cm.deliveryComposition(deliveryComposition.toComscore());
      if(deliveryAdvertisementCapability != null) cm.deliveryAdvertisementCapability(deliveryAdvertisementCapability.toComscore());
      if(mediaFormat != null) cm.mediaFormat(mediaFormat.toComscore());
      if(distributionModel != null) cm.distributionModel(distributionModel.toComscore());
      if(playlistTitle != null) cm.playlistTitle(playlistTitle);
      if(totalSegments != null) cm.totalSegments(totalSegments);
      if(clipUrl != null) cm.clipUrl(clipUrl);
      if(videoDimension != null) cm.videoDimensions(videoDimension.getWidth(),videoDimension.getHeight());
      if(!customLabels.isEmpty()){
        cm.customLabels(customLabels);
      }
      return cm.build();
    }

  public long getLength() {
      return length;
  }

  /**
     * The builder for {@link ComscoreMetaData}.
     */
    public static class Builder {
      private ComscoreMediaType mediaType;
      private String uniqueId;
      private long length;
      private String c3 = null;
      private String c4 = null;
      private String c6 = null;
      private String stationTitle;
      private String stationCode = null;
      private String networkAffiliate = null;
      private String publisherName = null;
      private String programTitle;
      private String programId = null;
      private String episodeTitle;
      private String episodeId = null;
      private String episodeSeasonNumber = null;
      private String episodeNumber = null;
      private String genreName;
      private String genreId = null;
      private Boolean carryTvAdvertisementLoad = null;
      private Boolean classifyAsCompleteEpisode = null;
      private ComscoreDate dateOfProduction = null;
      private ComscoreTime timeOfProduction = null;
      private ComscoreDate dateOfTvAiring = null;
      private ComscoreTime timeOfTvAiring = null;
      private ComscoreDate dateOfDigitalAiring = null;
      private ComscoreTime timeOfDigitalAiring = null;
      private ComscoreFeedType feedType = null;
      private Boolean classifyAsAudioStream;
      private ComscoreDeliveryMode deliveryMode = null;
      private ComscoreDeliverySubscriptionType deliverySubscriptionType = null;
      private ComscoreDeliveryComposition deliveryComposition = null;
      private ComscoreDeliveryAdvertisementCapability deliveryAdvertisementCapability = null;
      private ComscoreMediaFormat mediaFormat = null;
      private ComscoreDistributionModel distributionModel = null;
      private String playlistTitle = null;
      private Integer totalSegments = null;
      private String clipUrl = null;
      private ComscoreDimension videoDimension = null;
      private Map<String,String> customLabels = null;

        @NonNull
        private static ComscoreMetaData createComscoreMetadata(ComscoreMediaType mediaType, String uniqueId, long length, String c3, String c4, String c6, String stationTitle, String stationCode, String networkAffiliate, String publisherName, String programTitle, String programId, String episodeTitle, String episodeId, String episodeSeasonNumber, String episodeNumber, String genreName, String genreId, Boolean carryTvAdvertisementLoad, Boolean classifyAsCompleteEpisode, ComscoreDate dateOfProduction, ComscoreTime timeOfProduction, ComscoreDate dateOfTvAiring, ComscoreTime timeOfTvAiring, ComscoreDate dateOfDigitalAiring, ComscoreTime timeOfDigitalAiring, ComscoreFeedType feedType, Boolean classifyAsAudioStream, ComscoreDeliveryMode deliveryMode, ComscoreDeliverySubscriptionType deliverySubscriptionType, ComscoreDeliveryComposition deliveryComposition, ComscoreDeliveryAdvertisementCapability deliveryAdvertisementCapability, ComscoreMediaFormat mediaFormat, ComscoreDistributionModel distributionModel, String playlistTitle, Integer totalSegments, String clipUrl, ComscoreDimension videoDimension, Map<String,String> customLabels) {
            return new ComscoreMetaData(mediaType,uniqueId,length,c3,c4,c6,stationTitle,stationCode,networkAffiliate,publisherName,programTitle,programId,episodeTitle,episodeId,episodeSeasonNumber,episodeNumber,genreName,genreId,carryTvAdvertisementLoad,classifyAsCompleteEpisode,dateOfProduction,timeOfProduction,dateOfTvAiring,timeOfTvAiring,dateOfDigitalAiring,timeOfDigitalAiring,feedType,classifyAsAudioStream,deliveryMode,deliverySubscriptionType,deliveryComposition,deliveryAdvertisementCapability,mediaFormat,distributionModel,playlistTitle,totalSegments,clipUrl,videoDimension,customLabels);
        }

        /**
         * Creates a builder for a default Comscore metadata object.
         *
         * @remarks
         * <pre>
         *     - This configuration does not contain any sources.
         * </pre>
         */
        @NonNull
        public static Builder comscoreMetadata() {
            return new Builder();
        }

      public Builder mediaType(@NonNull ComscoreMediaType mediaType) {
        this.mediaType = mediaType;
        return this;
      }
      public Builder uniqueId(@NonNull String uniqueId) {
        this.uniqueId = uniqueId;
        return this;
      }
      public Builder length(@NonNull long length) {
        this.length = length;
        return this;
      }
      public Builder c3(@NonNull String c3) {
        this.c3 = c3;
        return this;
      }
      public Builder c4(@NonNull String c4) {
        this.c4 = c4;
        return this;
      }
      public Builder c6(@NonNull String c6) {
        this.c6 = c6;
        return this;
      }
      public Builder stationTitle(@NonNull String stationTitle) {
        this.stationTitle = stationTitle;
        return this;
      }
      public Builder stationCode(@NonNull String stationCode) {
        this.stationCode = stationCode;
        return this;
      }
      public Builder networkAffiliate(@NonNull String networkAffiliate) {
        this.networkAffiliate = networkAffiliate;
        return this;
      }
      public Builder publisherName(@NonNull String publisherName) {
        this.publisherName = publisherName;
        return this;
      }
      public Builder programTitle(@NonNull String programTitle) {
        this.programTitle = programTitle;
        return this;
      }
      public Builder programId(@NonNull String programId) {
        this.programId = programId;
        return this;
      }
      public Builder episodeTitle(@NonNull String episodeTitle) {
        this.episodeTitle = episodeTitle;
        return this;
      }
      public Builder episodeId(@NonNull String episodeId) {
        this.episodeId = episodeId;
        return this;
      }
      public Builder episodeSeasonNumber(@NonNull String episodeSeasonNumber) {
        this.episodeSeasonNumber = episodeSeasonNumber;
        return this;
      }
      public Builder episodeNumber(@NonNull String episodeNumber) {
        this.episodeNumber = episodeNumber;
        return this;
      }
      public Builder genreName(@NonNull String genreName) {
        this.genreName = genreName;
        return this;
      }
      public Builder genreId(@NonNull String genreId) {
        this.genreId = genreId;
        return this;
      }
      public Builder carryTvAdvertisementLoad(@NonNull Boolean carryTvAdvertisementLoad) {
        this.carryTvAdvertisementLoad = carryTvAdvertisementLoad;
        return this;
      }
      public Builder classifyAsCompleteEpisode(@NonNull Boolean classifyAsCompleteEpisode) {
        this.classifyAsCompleteEpisode = classifyAsCompleteEpisode;
        return this;
      }
      public Builder dateOfProduction(@NonNull ComscoreDate dateOfProduction) {
        this.dateOfProduction = dateOfProduction;
        return this;
      }
      public Builder timeOfProduction(@NonNull ComscoreTime timeOfProduction) {
        this.timeOfProduction = timeOfProduction;
        return this;
      }
      public Builder dateOfTvAiring(@NonNull ComscoreDate dateOfTvAiring) {
        this.dateOfTvAiring = dateOfTvAiring;
        return this;
      }
      public Builder timeOfTvAiring(@NonNull ComscoreTime timeOfTvAiring) {
        this.timeOfTvAiring = timeOfTvAiring;
        return this;
      }
      public Builder dateOfDigitalAiring(@NonNull ComscoreDate dateOfDigitalAiring) {
        this.dateOfDigitalAiring = dateOfDigitalAiring;
        return this;
      }
      public Builder timeOfDigitalAiring(@NonNull ComscoreTime timeOfDigitalAiring) {
        this.timeOfDigitalAiring = timeOfDigitalAiring;
        return this;
      }
      public Builder feedType(@NonNull ComscoreFeedType feedType) {
        this.feedType = feedType;
        return this;
      }
      public Builder classifyAsAudioStream(@NonNull Boolean classifyAsAudioStream) {
        this.classifyAsAudioStream = classifyAsAudioStream;
        return this;
      }
      public Builder deliveryMode(@NonNull ComscoreDeliveryMode deliveryMode) {
        this.deliveryMode = deliveryMode;
        return this;
      }
      public Builder deliverySubscriptionType(@NonNull ComscoreDeliverySubscriptionType deliverySubscriptionType) {
        this.deliverySubscriptionType = deliverySubscriptionType;
        return this;
      }
      public Builder deliveryComposition(@NonNull ComscoreDeliveryComposition deliveryComposition) {
        this.deliveryComposition = deliveryComposition;
        return this;
      }
      public Builder deliveryAdvertisementCapability(@NonNull ComscoreDeliveryAdvertisementCapability deliveryAdvertisementCapability) {
        this.deliveryAdvertisementCapability = deliveryAdvertisementCapability;
        return this;
      }
      public Builder mediaFormat(@NonNull ComscoreMediaFormat mediaFormat) {
        this.mediaFormat = mediaFormat;
        return this;
      }
      public Builder distributionModel(@NonNull ComscoreDistributionModel distributionModel) {
        this.distributionModel = distributionModel;
        return this;
      }
      public Builder playlistTitle(@NonNull String playlistTitle) {
        this.playlistTitle = playlistTitle;
        return this;
      }
      public Builder totalSegments(@NonNull Integer totalSegments) {
        this.totalSegments = totalSegments;
        return this;
      }
      public Builder clipUrl(@NonNull String clipUrl) {
        this.clipUrl = clipUrl;
        return this;
      }
      public Builder videoDimension(@NonNull ComscoreDimension videoDimension) {
        this.videoDimension = videoDimension;
        return this;
      }
      public Builder customLabels(@NonNull Map<String,String> customLabels) {
        this.customLabels = customLabels;
        return this;
      }

      public ComscoreMetaData build() {
          return createComscoreMetadata(mediaType,uniqueId,length,c3,c4,c6,stationTitle,stationCode,networkAffiliate,publisherName,programTitle,programId,episodeTitle,episodeId,episodeSeasonNumber,episodeNumber,genreName,genreId,carryTvAdvertisementLoad,classifyAsCompleteEpisode,dateOfProduction,timeOfProduction,dateOfTvAiring,timeOfTvAiring,dateOfDigitalAiring,timeOfDigitalAiring,feedType,classifyAsAudioStream,deliveryMode,deliverySubscriptionType,deliveryComposition,deliveryAdvertisementCapability,mediaFormat,distributionModel,playlistTitle,totalSegments,clipUrl,videoDimension,customLabels);
      }

    }
}

File: comscore/android/src/main/java/com/theoplayercomscore/integration/ComscoreMetaData.kt
package com.theoplayercomscore.integration

import com.comscore.streaming.ContentMetadata

class ComscoreMetaData(
  private val mediaType: ComscoreMediaType,
  private val uniqueId: String?,
  val length: Long,
  private val c3: String?,
  private val c4: String?,
  private val c6: String?,
  private val stationTitle: String?,
  private val stationCode: String?,
  private val networkAffiliate: String?,
  private val publisherName: String?,
  private val programTitle: String?,
  private val programId: String?,
  private val episodeTitle: String?,
  private val episodeId: String?,
  private val episodeSeasonNumber: String?,
  private val episodeNumber: String?,
  private val genreName: String?,
  private val genreId: String?,
  private val carryTvAdvertisementLoad: Boolean?,
  private val classifyAsCompleteEpisode: Boolean?,
  private val dateOfProduction: ComscoreDate?,
  private val timeOfProduction: ComscoreTime?,
  private val dateOfTvAiring: ComscoreDate?,
  private val timeOfTvAiring: ComscoreTime?,
  private val dateOfDigitalAiring: ComscoreDate?,
  private val timeOfDigitalAiring: ComscoreTime?,
  private val feedType: ComscoreFeedType?,
  private val classifyAsAudioStream: Boolean,
  private val deliveryMode: ComscoreDeliveryMode?,
  private val deliverySubscriptionType: ComscoreDeliverySubscriptionType?,
  private val deliveryComposition: ComscoreDeliveryComposition?,
  private val deliveryAdvertisementCapability: ComscoreDeliveryAdvertisementCapability?,
  private val mediaFormat: ComscoreMediaFormat?,
  private val distributionModel: ComscoreDistributionModel?,
  private val playlistTitle: String?,
  private val totalSegments: Int?,
  private val clipUrl: String?,
  private val videoDimension: ComscoreDimension?,
  private val customLabels: Map<String, String>
) {

  fun toComscoreContentMetadata(): ContentMetadata {
    val cm = ContentMetadata.Builder()
    cm.mediaType(mediaType.toComscore())
    cm.uniqueId(uniqueId)
    cm.length(length)
    if (c3 != null) cm.dictionaryClassificationC3(c3)
    if (c4 != null) cm.dictionaryClassificationC4(c4)
    if (c6 != null) cm.dictionaryClassificationC6(c6)
    cm.stationTitle(stationTitle)
    if (stationCode != null) cm.stationCode(stationCode)
    if (networkAffiliate != null) cm.networkAffiliate(networkAffiliate)
    if (publisherName != null) cm.publisherName(publisherName)
    cm.programTitle(programTitle)
    if (programId != null) cm.programId(programId)
    cm.episodeTitle(episodeTitle)
    if (episodeId != null) cm.episodeId(episodeId)
    if (episodeSeasonNumber != null) cm.episodeSeasonNumber(episodeSeasonNumber)
    if (episodeNumber != null) cm.episodeNumber(episodeNumber)
    cm.genreName(genreName)
    if (genreId != null) cm.genreId(genreId)
    if (carryTvAdvertisementLoad != null) cm.carryTvAdvertisementLoad(carryTvAdvertisementLoad)
    if (classifyAsCompleteEpisode != null) cm.classifyAsCompleteEpisode(
      classifyAsCompleteEpisode
    )
    if (dateOfProduction != null) cm.dateOfProduction(
      dateOfProduction.year,
      dateOfProduction.month,
      dateOfProduction.day
    )
    if (timeOfProduction != null) cm.timeOfProduction(
      timeOfProduction.hours,
      timeOfProduction.minutes
    )
    if (dateOfTvAiring != null) cm.dateOfTvAiring(
      dateOfTvAiring.year,
      dateOfTvAiring.month,
      dateOfTvAiring.day
    )
    if (timeOfTvAiring != null) cm.timeOfTvAiring(timeOfTvAiring.hours, timeOfTvAiring.minutes)
    if (dateOfDigitalAiring != null) cm.dateOfDigitalAiring(
      dateOfDigitalAiring.year,
      dateOfDigitalAiring.month,
      dateOfDigitalAiring.day
    )
    if (timeOfDigitalAiring != null) cm.timeOfDigitalAiring(
      timeOfDigitalAiring.hours,
      timeOfDigitalAiring.minutes
    )
    if (feedType != null) cm.feedType(feedType.toComscore())
    cm.classifyAsAudioStream(classifyAsAudioStream)
    if (deliveryMode != null) cm.deliveryMode(deliveryMode.toComscore())
    if (deliverySubscriptionType != null) cm.deliverySubscriptionType(deliverySubscriptionType.toComscore())
    if (deliveryComposition != null) cm.deliveryComposition(deliveryComposition.toComscore())
    if (deliveryAdvertisementCapability != null) cm.deliveryAdvertisementCapability(
      deliveryAdvertisementCapability.toComscore()
    )
    if (mediaFormat != null) cm.mediaFormat(mediaFormat.toComscore())
    if (distributionModel != null) cm.distributionModel(distributionModel.toComscore())
    if (playlistTitle != null) cm.playlistTitle(playlistTitle)
    if (totalSegments != null) cm.totalSegments(totalSegments)
    if (clipUrl != null) cm.clipUrl(clipUrl)
    if (videoDimension != null) cm.videoDimensions(videoDimension.width, videoDimension.height)
    if (customLabels.isNotEmpty()) {
      cm.customLabels(customLabels)
    }
    return cm.build()
  }

  /**
   * The builder for [ComscoreMetaData].
   */
  class Builder {
    private var mediaType: ComscoreMediaType = ComscoreMediaType.OTHER
    private var uniqueId: String? = null
    private var length: Long = 0
    private var c3: String? = null
    private var c4: String? = null
    private var c6: String? = null
    private var stationTitle: String? = null
    private var stationCode: String? = null
    private var networkAffiliate: String? = null
    private var publisherName: String? = null
    private var programTitle: String? = null
    private var programId: String? = null
    private var episodeTitle: String? = null
    private var episodeId: String? = null
    private var episodeSeasonNumber: String? = null
    private var episodeNumber: String? = null
    private var genreName: String? = null
    private var genreId: String? = null
    private var carryTvAdvertisementLoad: Boolean? = null
    private var classifyAsCompleteEpisode: Boolean? = null
    private var dateOfProduction: ComscoreDate? = null
    private var timeOfProduction: ComscoreTime? = null
    private var dateOfTvAiring: ComscoreDate? = null
    private var timeOfTvAiring: ComscoreTime? = null
    private var dateOfDigitalAiring: ComscoreDate? = null
    private var timeOfDigitalAiring: ComscoreTime? = null
    private var feedType: ComscoreFeedType? = null
    private var classifyAsAudioStream: Boolean = false
    private var deliveryMode: ComscoreDeliveryMode? = null
    private var deliverySubscriptionType: ComscoreDeliverySubscriptionType? = null
    private var deliveryComposition: ComscoreDeliveryComposition? = null
    private var deliveryAdvertisementCapability: ComscoreDeliveryAdvertisementCapability? = null
    private var mediaFormat: ComscoreMediaFormat? = null
    private var distributionModel: ComscoreDistributionModel? = null
    private var playlistTitle: String? = null
    private var totalSegments: Int? = null
    private var clipUrl: String? = null
    private var videoDimension: ComscoreDimension? = null
    private var customLabels: Map<String, String> = mapOf()

    fun mediaType(mediaType: ComscoreMediaType): Builder {
      this.mediaType = mediaType
      return this
    }

    fun uniqueId(uniqueId: String): Builder {
      this.uniqueId = uniqueId
      return this
    }

    fun length(length: Long): Builder {
      this.length = length
      return this
    }

    fun c3(c3: String): Builder {
      this.c3 = c3
      return this
    }

    fun c4(c4: String): Builder {
      this.c4 = c4
      return this
    }

    fun c6(c6: String): Builder {
      this.c6 = c6
      return this
    }

    fun stationTitle(stationTitle: String): Builder {
      this.stationTitle = stationTitle
      return this
    }

    fun stationCode(stationCode: String): Builder {
      this.stationCode = stationCode
      return this
    }

    fun networkAffiliate(networkAffiliate: String): Builder {
      this.networkAffiliate = networkAffiliate
      return this
    }

    fun publisherName(publisherName: String): Builder {
      this.publisherName = publisherName
      return this
    }

    fun programTitle(programTitle: String): Builder {
      this.programTitle = programTitle
      return this
    }

    fun programId(programId: String): Builder {
      this.programId = programId
      return this
    }

    fun episodeTitle(episodeTitle: String): Builder {
      this.episodeTitle = episodeTitle
      return this
    }

    fun episodeId(episodeId: String): Builder {
      this.episodeId = episodeId
      return this
    }

    fun episodeSeasonNumber(episodeSeasonNumber: String): Builder {
      this.episodeSeasonNumber = episodeSeasonNumber
      return this
    }

    fun episodeNumber(episodeNumber: String): Builder {
      this.episodeNumber = episodeNumber
      return this
    }

    fun genreName(genreName: String): Builder {
      this.genreName = genreName
      return this
    }

    fun genreId(genreId: String): Builder {
      this.genreId = genreId
      return this
    }

    fun carryTvAdvertisementLoad(carryTvAdvertisementLoad: Boolean): Builder {
      this.carryTvAdvertisementLoad = carryTvAdvertisementLoad
      return this
    }

    fun classifyAsCompleteEpisode(classifyAsCompleteEpisode: Boolean): Builder {
      this.classifyAsCompleteEpisode = classifyAsCompleteEpisode
      return this
    }

    fun dateOfProduction(dateOfProduction: ComscoreDate): Builder {
      this.dateOfProduction = dateOfProduction
      return this
    }

    fun timeOfProduction(timeOfProduction: ComscoreTime): Builder {
      this.timeOfProduction = timeOfProduction
      return this
    }

    fun dateOfTvAiring(dateOfTvAiring: ComscoreDate): Builder {
      this.dateOfTvAiring = dateOfTvAiring
      return this
    }

    fun timeOfTvAiring(timeOfTvAiring: ComscoreTime): Builder {
      this.timeOfTvAiring = timeOfTvAiring
      return this
    }

    fun dateOfDigitalAiring(dateOfDigitalAiring: ComscoreDate): Builder {
      this.dateOfDigitalAiring = dateOfDigitalAiring
      return this
    }

    fun timeOfDigitalAiring(timeOfDigitalAiring: ComscoreTime): Builder {
      this.timeOfDigitalAiring = timeOfDigitalAiring
      return this
    }

    fun feedType(feedType: ComscoreFeedType): Builder {
      this.feedType = feedType
      return this
    }

    fun classifyAsAudioStream(classifyAsAudioStream: Boolean): Builder {
      this.classifyAsAudioStream = classifyAsAudioStream
      return this
    }

    fun deliveryMode(deliveryMode: ComscoreDeliveryMode): Builder {
      this.deliveryMode = deliveryMode
      return this
    }

    fun deliverySubscriptionType(deliverySubscriptionType: ComscoreDeliverySubscriptionType): Builder {
      this.deliverySubscriptionType = deliverySubscriptionType
      return this
    }

    fun deliveryComposition(deliveryComposition: ComscoreDeliveryComposition): Builder {
      this.deliveryComposition = deliveryComposition
      return this
    }

    fun deliveryAdvertisementCapability(deliveryAdvertisementCapability: ComscoreDeliveryAdvertisementCapability): Builder {
      this.deliveryAdvertisementCapability = deliveryAdvertisementCapability
      return this
    }

    fun mediaFormat(mediaFormat: ComscoreMediaFormat): Builder {
      this.mediaFormat = mediaFormat
      return this
    }

    fun distributionModel(distributionModel: ComscoreDistributionModel): Builder {
      this.distributionModel = distributionModel
      return this
    }

    fun playlistTitle(playlistTitle: String): Builder {
      this.playlistTitle = playlistTitle
      return this
    }

    fun totalSegments(totalSegments: Int): Builder {
      this.totalSegments = totalSegments
      return this
    }

    fun clipUrl(clipUrl: String): Builder {
      this.clipUrl = clipUrl
      return this
    }

    fun videoDimension(videoDimension: ComscoreDimension): Builder {
      this.videoDimension = videoDimension
      return this
    }

    fun customLabels(customLabels: Map<String, String>): Builder {
      this.customLabels = customLabels
      return this
    }

    fun build(): ComscoreMetaData {
      return createComscoreMetadata(
        mediaType,
        uniqueId,
        length,
        c3,
        c4,
        c6,
        stationTitle,
        stationCode,
        networkAffiliate,
        publisherName,
        programTitle,
        programId,
        episodeTitle,
        episodeId,
        episodeSeasonNumber,
        episodeNumber,
        genreName,
        genreId,
        carryTvAdvertisementLoad,
        classifyAsCompleteEpisode,
        dateOfProduction,
        timeOfProduction,
        dateOfTvAiring,
        timeOfTvAiring,
        dateOfDigitalAiring,
        timeOfDigitalAiring,
        feedType,
        classifyAsAudioStream,
        deliveryMode,
        deliverySubscriptionType,
        deliveryComposition,
        deliveryAdvertisementCapability,
        mediaFormat,
        distributionModel,
        playlistTitle,
        totalSegments,
        clipUrl,
        videoDimension,
        customLabels
      )
    }

    companion object {
      private fun createComscoreMetadata(
        mediaType: ComscoreMediaType,
        uniqueId: String?,
        length: Long,
        c3: String?,
        c4: String?,
        c6: String?,
        stationTitle: String?,
        stationCode: String?,
        networkAffiliate: String?,
        publisherName: String?,
        programTitle: String?,
        programId: String?,
        episodeTitle: String?,
        episodeId: String?,
        episodeSeasonNumber: String?,
        episodeNumber: String?,
        genreName: String?,
        genreId: String?,
        carryTvAdvertisementLoad: Boolean?,
        classifyAsCompleteEpisode: Boolean?,
        dateOfProduction: ComscoreDate?,
        timeOfProduction: ComscoreTime?,
        dateOfTvAiring: ComscoreDate?,
        timeOfTvAiring: ComscoreTime?,
        dateOfDigitalAiring: ComscoreDate?,
        timeOfDigitalAiring: ComscoreTime?,
        feedType: ComscoreFeedType?,
        classifyAsAudioStream: Boolean,
        deliveryMode: ComscoreDeliveryMode?,
        deliverySubscriptionType: ComscoreDeliverySubscriptionType?,
        deliveryComposition: ComscoreDeliveryComposition?,
        deliveryAdvertisementCapability: ComscoreDeliveryAdvertisementCapability?,
        mediaFormat: ComscoreMediaFormat?,
        distributionModel: ComscoreDistributionModel?,
        playlistTitle: String?,
        totalSegments: Int?,
        clipUrl: String?,
        videoDimension: ComscoreDimension?,
        customLabels: Map<String, String>
      ): ComscoreMetaData {
        return ComscoreMetaData(
          mediaType,
          uniqueId,
          length,
          c3,
          c4,
          c6,
          stationTitle,
          stationCode,
          networkAffiliate,
          publisherName,
          programTitle,
          programId,
          episodeTitle,
          episodeId,
          episodeSeasonNumber,
          episodeNumber,
          genreName,
          genreId,
          carryTvAdvertisementLoad,
          classifyAsCompleteEpisode,
          dateOfProduction,
          timeOfProduction,
          dateOfTvAiring,
          timeOfTvAiring,
          dateOfDigitalAiring,
          timeOfDigitalAiring,
          feedType,
          classifyAsAudioStream,
          deliveryMode,
          deliverySubscriptionType,
          deliveryComposition,
          deliveryAdvertisementCapability,
          mediaFormat,
          distributionModel,
          playlistTitle,
          totalSegments,
          clipUrl,
          videoDimension,
          customLabels
        )
      }

      /**
       * Creates a builder for a default Comscore metadata object.
       *
       * <pre>
       * - This configuration does not contain any sources.
      </pre> *
       */
      fun comscoreMetadata(): Builder {
        return Builder()
      }
    }
  }
}
